src/Conversion/ONNXToMhlo/NN/Conv.cpp

```diff
@@ -0,0 +1,133 @@
+#include "src/Support/TypeUtilities.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+struct ONNXConvOpLoweringToMhlo : public ConversionPattern {
+  ONNXConvOpLoweringToMhlo(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXConvOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    ONNXConvOpAdaptor operandAdaptor(operands);
+    ONNXConvOp convOp = llvm::dyn_cast<ONNXConvOp>(op);
+    Location loc = op->getLoc();
+
+    ONNXConvOpShapeHelper shapeHelper(&convOp);
+    LogicalResult shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    llvm::SmallVector<IndexExpr, 2> kernelShape = shapeHelper.kernelShape;
+    llvm::SmallVector<int64_t, 2> strides = shapeHelper.strides;
+    llvm::SmallVector<int64_t, 2> dilations = shapeHelper.dilations;
+    DimsExpr outputDims = shapeHelper.dimsForOutput();
+    int outputRank = shapeHelper.dimsForOutput().size();
+
+    Value inputOperand = operandAdaptor.X();
+    Value filterOperand = operandAdaptor.W();
+    Value biasOperand = operandAdaptor.B();
+    bool hasBias = !biasOperand.getType().isa<NoneType>();
+    int64_t groupNum = convOp.group();
+
+    assert(isRankedShapedType(inputOperand.getType()) &&
+           "Expected Ranked ShapedType");
+    ShapedType inputType = inputOperand.getType().cast<ShapedType>();
+    llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
+    Type outputType = *op->result_type_begin();
+    // Onnx Input is NCHW
+    int64_t spatialOffset = 2;
+    int64_t rank = inputType.getRank();
+    int64_t kernelSize = kernelShape.size();
+
+    SmallVector<int64_t> inputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < rank; i++) {
+      inputSpatialDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> kernelDimensions;
+    for (int64_t i = spatialOffset; i < spatialOffset + kernelSize; i++) {
+      kernelDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> outputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < outputRank; i++) {
+      outputSpatialDimensions.push_back(i);
+    }
+
+    // paddings
+    DimsExpr pads = shapeHelper.pads;
+    llvm::StringRef padding = convOp.auto_pad();
+    int64_t spatialRank = rank - spatialOffset;
+    SmallVector<int64_t> flattenPaddings;
+    bool needPadding = (padding == "NOTSET");
+    for (int64_t i = 0; i < spatialRank; i++) {
+      if (!needPadding) {
+        flattenPaddings.push_back(pads[i].getLiteral());
+        flattenPaddings.push_back(pads[i + spatialRank].getLiteral());
+      } else {
+        int64_t kdTerm = (kernelShape[i].getLiteral() - 1) * dilations[i] + 1;
+        int64_t padFront = pads[i].getLiteral();
+        int64_t padBack =
+            (outputDims[i + spatialOffset].getLiteral() - 1) * strides[i] +
```

Comment:
do we know that all these things we call getLiteral() on are literals?

Reply:
The current conv only supports where padding and kernel size are known. We will add support for dynamic shape.

Reply:
I see

I suggest you add assertions to make it clear that pads, kernelShape, outputDims, and inputShape must all be known (unless this is already checked elsewhere)

Reply:
My recollection is that we support dynamic images sizes, but expect the filter size to be known at compile time...
I will try to add some dyn size testing if its not already there.

Reply:
Confirmed, in test/modellib/ConvModel and test/numerical/TestConv we are building/testing 2D convolutions that are dynamic in each of the NCHW dimensions.